### PR TITLE
Add notification for third party services

### PR DESCRIPTION
**Pull request by joshpme**
This will allow downstream systems to be notified when revisions are created in indico specifically for JACoW

It is fire and forget to prevent any outage affecting indico uptime.

The hope is that will enable a downstream system "e.g." catscan to perform a scan on a change, rather than polling.

Also first python PR ever, so I welcome any feedback.

If you are happy to merge and deploy the change, we can discuss the URL / Token on slack

### CODE REVIEW

File: openreferee_server/notify.py
Comment (ThiefMaster) [inline openreferee_server/notify.py:17]: No. Don't do that. At least print the exception and traceback if you decide to ignore it. But don't silently drop *any* exception. Also, be more specific (request failing, possibly a timeout as well) and don't catch just about anything. If there's let's say a TypeError (just an example) that'd be a bug and not something you want to silently ignore.

File: openreferee_server/app.py
Comment (ThiefMaster) [inline openreferee_server/app.py:29]: Use `os.environ[...]` if you expect this to always be present

File: openreferee_server/notify.py
Comment (ThiefMaster) [inline openreferee_server/notify.py:43]: I'm not sure if it makes much sense for this whole thing to be implemented using the flask extension pattern, but regardless of that: This check is not necessary because in recent Flask versions the `context.extension` dict is already guaranteed to exist.

File: openreferee_server/server.py
Comment (ThiefMaster) [inline openreferee_server/server.py:49]: You `.get()` it but if it's `None`, calling `notify()` on it will fail. so use `context.extensions['notifier']` here or handle the case where `get_notifier()` returns `None` below

File: openreferee_server/notify.py
Comment (ThiefMaster) [inline openreferee_server/notify.py:15]: Why is this a get request with a payload? Yes, technically a GET request can have a body, but it's incredibly uncommon. Use POST or similar - for *triggering* some action based on a notification this makes more sense anyway. GET should never perform an action.

File: openreferee_server/notify.py
Comment (ThiefMaster) [inline openreferee_server/notify.py:26]: I wonder... is all this extra complexity really needed? To me it sounds like the whole code in here could be done with a simple

```python
def notify(payload):
    sess = ...
    try:
        sess.post(..., timeout=5)
    except ...:
        pass  # log it
```

no threads, no flask extension, no context etc needed ;)

Comment (joshpme) [reply openreferee_server/notify.py:26]: Sure I thought I might be over-complicating this, but I thought I'd be adding the notify to each of the endpoints, or maybe extending the logic out if required.

The reason why I use an extension, I needed to load in the environmental config for the notify URL. So I figured I'd need an extension to load this in? Should I just load from process.env from anywhere?

The threading is there as I wanted this to me fire and forget so there would be no slow down to this service if our downstream system was offline or slow to response.

The two reasons I'm using a session is that I am able to config the authentication at a parent level using the method you already have, and depending on the traffic in this service, having a common session between requests reduces http overhead. I'm not sure how it cleans itself up though, so I'm happy to simply this if you do not believe there is much traffic?

File: openreferee_server/notify.py
Comment (joshpme) [reply openreferee_server/notify.py:15]: Just a mistake on my part, I will change this to a post.

File: openreferee_server/notify.py
Comment (ThiefMaster) [reply openreferee_server/notify.py:26]: > The reason why I use an extension, I needed to load in the environmental config for the notify URL. So I figured I'd need an extension to load this in? Should I just load from process.env from anywhere?

Yes, you can read env vars anywhere. Or even put it in the `app.config` dict and then access it from there later.

> The threading is there as I wanted this to me fire and forget so there would be no slow down to this service if our downstream system was offline or slow to response.

Fair enough, I don't know how reliable your downstream is. Personally I'd probably just use a low-enough timeout and then only log request-related errors so they don't break the flow.

> The two reasons I'm using a session is that I am able to config the authentication at a parent level using the method you already have, and depending on the traffic in this service, having a common session between requests reduces http overhead. I'm not sure how it cleans itself up though, so I'm happy to simply this if you do not believe there is much traffic?

Actually, I'm not sure if the session actually keeps connections alive etc. or not. Using it for the token is a good idea though!

File: openreferee_server/app.py
Comment (joshpme) [reply openreferee_server/app.py:29]: This will throw an exception if the keys are not present, where as get will soft return None

I explicitly handle none, plus as these variables can be figured with flask config instead of environmental variables if that is needed later.

Is there a reason why you prefer to use a dict lookup over get?

Comment (ThiefMaster) [reply openreferee_server/app.py:29]: I generally prefer to fail early: If required config keys are missing, I want it to fail at startup time. That's something I see immediately because the app is clearly broken.

If the missing value leaks into the config and it only fails later when trying to send a request, there's a good chance I won't notice it until a real request from a user's action arrives.

That said, I think it makes sense to simply disable notifications if these settings are missing. I'd do that by not initializing the extension and silently doing nothing later when the extension is not on the app context (based on your current architecture). Without the extension structure I'd simply return early in the function to send a notification if the config values are None,

File: openreferee_server/notify.py
Comment (ThiefMaster) [inline openreferee_server/notify.py:44]: The `init_app` argument is the `app` itself, not an application `context`

File: openreferee_server/notify.py
Comment (joshpme) [reply openreferee_server/notify.py:33]: Is this call to the microservice blocking, as in, will buttons take longer to load if we put in a downstream call? 

I expect that the (yet-to-be-built) downstream service may be hosted on the other side of the planet, also I really didn't want to be responsible for an extra 2-5 second load times on some buttons, or prematurely timeout cause of distance.

That being said, I'm also no python guru, so maybe threads are not the best tool for the job here (due to leaky threads or contention issues)? So if you feel strongly about this I will just change it. 

Also regarding the use of an extension, I was just following a pattern I'm familiar with from other (non-python) frameworks, so if you want me to simplify it, no problem, I'm just not sure if you comment was more of an observation or a suggestion for change.

File: openreferee_server/app.py
Comment (joshpme) [reply openreferee_server/app.py:29]: Yeah, fair. I'll refactor this now then.

**Comment by joshpme** [comment]
@ThiefMaster is there anything you still need me to do on this PR before it can be merged?

File: openreferee_server/app.py
Comment (ThiefMaster) [inline openreferee_server/app.py:8]: please run `isort` on the files

File: openreferee_server/notify.py
Comment (ThiefMaster) [inline openreferee_server/notify.py:18]: I don't think returning the result here makes sense. It's always None

File: openreferee_server/notify.py
Comment (ThiefMaster) [inline openreferee_server/notify.py:16]:
```suggestion
    def log_error(self, *args):
```

I don't think return type annotations are particularly useful for this, and this one was wrong as well: None of the logging methods actually return anything.

File: openreferee_server/notify.py
Comment (ThiefMaster) [inline openreferee_server/notify.py:33]:
```suggestion
        if not self.url:
```

File: openreferee_server/notify.py
Comment (ThiefMaster) [inline openreferee_server/notify.py:46]:
```suggestion
    if app.config['NOTIFY_URL']:
```

**Review by duartegalvao — CHANGES_REQUESTED** [review]
yo, hows it going? overall the PR lgtm, and from the few tests I did it didn't seem to break anything. I just have a couple of code style suggestions but thats it

cheers

File: openreferee_server/notify.py
Comment (duartegalvao) [reply openreferee_server/notify.py:56]: (just saw Adrian's comment, yeah theres no need for the `in` at all)

Comment (ThiefMaster) [reply openreferee_server/notify.py:56]: (if you accept this suggestion, please apply the suggestion from my review as well, the `in none, ''` check is quite weird)

File: openreferee_server/notify.py
Comment (duartegalvao) [reply openreferee_server/notify.py:39]: the buttons won't take longer to load since you're not doing anything to the `get_custom_revision_actions` endpoint, but still it's probably better to do it this way if you're unsure of the reliability of the service

**Review by duartegalvao — APPROVED** [review]
lgtm! might wanna get rid of that merge commit tho, we usually prefer rebasing
(not sure if this matters for this repo, I'll let @ThiefMaster decide)

**Comment by ThiefMaster** [comment]
I'll squash+merge anyway, so doesn't really matter.

**Pull request merged**
